tests/data/semparse/type_declarations/type_declaration_test.py

```diff
@@ -8,3 +11,57 @@ def test_basic_types_conflict_on_names(self):
         type_a = types.NamedBasicType("A")
         type_b = types.NamedBasicType("B")
         assert type_a.resolve(type_b) is None
+
+    def test_valid_actions(self):
+        type_r = types.NamedBasicType("R")
+        type_d = types.NamedBasicType("D")
+        type_e = types.NamedBasicType("E")
+        name_mapping = {'sample_function': 'F'}
+        # <e,<r,<d,r>>>
+        type_signatures = {'F': ComplexType(type_e, ComplexType(type_r, ComplexType(type_d, type_r)))}
+        basic_types = {type_r, type_d, type_e}
+        valid_actions = types.get_valid_actions(name_mapping, type_signatures, basic_types)
+        assert len(valid_actions) == 4
+        assert valid_actions["<e,<r,<d,r>>>"] == {"<e,<r,<d,r>>> -> sample_function"}
+        assert valid_actions["<r,<d,r>>"] == {"<r,<d,r>> -> [<e,<r,<d,r>>>, e]"}
+        assert valid_actions["<d,r>"] == {"<d,r> -> [<r,<d,r>>, r]"}
```

Comment:
Hardy har har...  https://www.youtube.com/watch?v=HcD1oA_s4-U - your type is `[<r,<d,r>>, r]`.  It's like you did that on purpose... =)

Reply:
:smile: